### PR TITLE
Add iterator for DeviceCollection as well as alternative default devices

### PR DESCRIPTION
**Pull request by 3tilley**
Is this what you imagined? Any strong feelings on `DeviceCollectionIter` instead of `DeviceIter`?

Also would I bump the version in this PR, or is that something you do prior to release?

### CODE REVIEW

**Comment by HEnquist** [comment]
This looks quite nice. But the DeviceCollectionIter should only borrow the DeviceCollection. Now it takes ownership so you won't be able to use the DeviceCollection again after iterating.
The borrow needs lifetime annotations, just copy those from alsa-rs.

The name of the iterator struct is not really important since this struct is mostly hidden from library users.

**Comment by 3tilley** [comment]
Thanks for your feedback, I really appreciate your guidance - changes made.

Is there a reason `alsa-rs` hasn't implemented `IntoIterator` apart from just not thinking it was a priority? I'm just interested as to whether there are corner cases where it's a bad idea, or if it's only an ergonomic positive.

**Comment by HEnquist** [comment]
Looks good, I'll take it for a spin as soon as I'm on a windows machine.

I think that IntoIter makes sense for things where it's obvious what the iterator will iterate over. It works perfectly for an array for example. It's also a good fit here for the Wasapi DeviceCollection. It will obviously iterate over devices. But an Alsa Ctl has a lot of functionality where nothing is clearly _the_ thing to iterate over.

**Comment by 3tilley** [comment]
That's a good find! Updated

File: src/api.rs [hunk 1]
```diff
@@ -93,6 +93,24 @@ impl fmt::Display for Direction {
     }
 }
 
+/// Audio direction, playback or capture.
```
Comment (HEnquist) [inline src/api.rs:96]: Copy-paste problem, please update :)
I think the link to the Microsoft docs for role would fit in better here than at `get_default_device()`.

Comment (3tilley) [reply src/api.rs]: Yes I agree

**Pull request merged**